V2 business

Alexvozhak opened the pull request:
## Summary

- Increased the maximum GateSeal duration from **14 to 21 days**.
- Allowed an initial expiry period of up to **3 years**.
- Introduced support for **DAO-controlled expiry extensions**:
  - The number of extensions is configurable (up to **5** times).
  - Each extension can prolong the seal by up to **6 months**.
- Added checks ensuring that a GateSeal can only be extended if:
  - It **has not been applied**,
  - It is **not expired**,
  - It **has remaining extensions** available.
- Extended the `GateSealFactory` to accept DAO configuration and extension parameters.
- Updated tests and documentation to reflect the new expiry and extension logic.

All technical and routine changes have been moved to a separate pull request (see [here](https://github.com/lidofinance/gate-seals/pull/34))

The design and implementation were inspired by [Gate Seals V2 specification](https://hackmd.io/@lido/gate-seals-v2).

Alexvozhak committed:
chore: Update dependencies and apply necessary project adjustments

Alexvozhak committed:
chore: Update poetry.lock and yarn.lock

Alexvozhak committed:
refactor: fix typo EIP 522 to EIP 5202

Alexvozhak committed:
chore: Bump Python to 3.10.13 and update poetry.lock

Alexvozhak committed:
fix: remove duplicate constant definition

Alexvozhak committed:
chore: remove .python-version file

Alexvozhak committed:
chore: update Node.js to v22 LTS

Alexvozhak committed:
docs: fix typo in README.md

Alexvozhak committed:
feat: implement GateSeal v2 business logic

Alexvozhak committed:
chore: update dependencies in yarn.lock

Alexvozhak committed:
docs: add Codex Agent instructions for setup, testing, and deployment

failingtwice commented on contracts/GateSeal.vy:
> GateSeals are only a temporary solution and will be deprecated in the future,
     as it is undesirable for the protocol to rely on a multisig. 
     
I guess this is no longer true, and gate seals are a permanent solution..

failingtwice replied:
```
event Sealed:
    gate_seal: address
    sealed_by: address
    sealed_for: uint256
    sealable: address
    sealed_at: uint256
```
Including the gate seal address in the event is redundant I think? You can't listen to this event without knowing the address of the contract

failingtwice replied:
```
# A unix epoch timestamp starting from which GateSeal is completely unusable
# and a new GateSeal will have to be set up. This timestamp will be changed
# upon sealing to expire GateSeal immediately which will revert any consecutive sealings.
```

Not sure which line this comment pertains to. Maybe move to a more appropriate place

failingtwice replied:
```
# The minimum allowed seal duration is 4 days. This is because it takes at least
# 3 days to pass and enact. Additionally, we want to include a 1-day padding.
MIN_SEAL_DURATION_DAYS: constant(uint256) = 4
```

This is no longer true. The DAO vote now lasts 5 days (3 days main phase, 2 days objection phase). Plus 1-day padding comes to a 6-day min duration.

failingtwice replied:
```
@notice converts a list of indexes into an error message to facilitate debugging
    @dev    The indexes in the error message are given in the descending order to avoid
            losing leading zeros when casting to string,
            e.g. [0, 2, 3, 6] -> "6320"
    @param _failed_indexes a list of sealable indexes that failed to seal 
```

Do we still need to this in the new Vyper version? I think they might have added a way to use dynamic strings as error message

Alexvozhak replied:
Yes, we still need to. Proof:
`curl -s "https://api.github.com/repos/vyperlang/vyper/compare/v0.3.7...v0.4.2"   | jq -r '.commits[].commit.message'   | grep -i dynamic | grep -i string || echo "Dynamic string support not found"`

failingtwice commented on contracts/GateSeal.vy:
```diff
@@ -23,8 +23,8 @@
 
-# The maximum GateSeal expiry duration is 1 year.
-MAX_EXPIRY_PERIOD_DAYS: constant(uint256) = 365
+# The maximum GateSeal expiry duration is 3 years.
+MAX_EXPIRY_PERIOD_DAYS: constant(uint256) = 365 * 3
 MAX_EXPIRY_PERIOD_SECONDS: constant(uint256) = SECONDS_PER_DAY * MAX_EXPIRY_PERIOD_DAYS
```
Consider grouping these together with the rest of the constants.

failingtwice commented on contracts/GateSeal.vy:
```diff
@@ -23,8 +23,8 @@
-# The maximum GateSeal expiry duration is 1 year.
-MAX_EXPIRY_PERIOD_DAYS: constant(uint256) = 365
+# The maximum GateSeal expiry duration is 3 years.
+MAX_EXPIRY_PERIOD_DAYS: constant(uint256) = 365 * 3
```
Since the duration can now span over multiple years (including leap years), we can be more precise by making it 365.25. Or maybe not, not too important I guess

Alexvozhak replied:
Good point. But since the cap is 3 years, the 0.75-day difference (≈64,800 seconds) doesn’t seem critical — and supporting non-integer durations would likely bring more complexity than practical benefit, IMO. I’d prefer to keep it as is

failingtwice commented on contracts/GateSeal.vy:
```diff
@@ -74,6 +74,11 @@ EXPIRY_MUST_BE_FUTURE: constant(String[39]) = "expiry timestamp: must be in the
 EXPIRY_EXCEEDS_MAX: constant(String[43]) = "expiry timestamp: exceeds max expiry period"
 SEALABLES_INCLUDES_ZERO: constant(String[32]) = "sealables: includes zero address"
 SEALABLES_INCLUDES_DUPLICATES: constant(String[30]) = "sealables: includes duplicates"
+DAO_ZERO_ADDRESS: constant(String[18]) = "dao: zero address"
+EXTENSION_DURATION_ZERO: constant(String[24]) = "extension duration: zero"
+SENDER_NOT_DAO: constant(String[17]) = "sender: not DAO"
+EXTENSIONS_EXHAUSTED: constant(String[22]) = "extensions: exhausted"
+GATE_SEAL_EXPIRED: constant(String[18]) = "gate seal: expired"
 
```
I think it's better to inline the errors, so errors are readable and accessible at the precise place they are thrown

failingtwice commented on contracts/GateSeal.vy:
```diff
@@ -86,33 +91,46 @@ SEALING_COMMITTEE: immutable(address)
+
 sealables: DynArray[address, MAX_SEALABLES]
 expiry_timestamp: uint256
+extensions_remaining: uint256
```
"Prolongations" could be more precise, since "extension" has more different meanings

failingtwice commented on contracts/GateSeal.vy:
```diff
@@ -86,33 +91,46 @@ SEALING_COMMITTEE: immutable(address)
 SEAL_DURATION_SECONDS: immutable(uint256)
 
+DAO: immutable(address)
+EXTENSION_DURATION_SECONDS: immutable(uint256)
```
Wouldn't it make more sense to decide on the extension duration at runtime? There may be situations (I imagine) where you don't want to prolong gate seal by too much.

failingtwice commented on contracts/GateSeal.vy:
```diff
@@ -139,12 +157,45 @@ def get_expiry_timestamp() -> uint256:
+    @notice Extend the GateSeal expiry.
+    @dev    Can be called only by DAO while the seal hasn't been used and not expired.
+    """
+    assert msg.sender == DAO, SENDER_NOT_DAO
```
Wouldn't it make more sense for the committee to do it? By shifting this responsibility to the committee, we can ensure that the committee is still in tact (e.g. no member lost their keys)

failingtwice commented on contracts/GateSeal.vy:
```diff
@@ -139,12 +157,45 @@ def get_expiry_timestamp() -> uint256:
+    @dev    Can be called only by DAO while the seal hasn't been used and not expired.
+    """
+    assert msg.sender == DAO, SENDER_NOT_DAO
+    assert self.expiry_timestamp > block.timestamp, GATE_SEAL_EXPIRED
```
```suggestion
    assert not self._is_expired(), GATE_SEAL_EXPIRED
```

failingtwice commented on contracts/GateSeal.vy:
```diff
@@ -139,12 +157,45 @@ def get_expiry_timestamp() -> uint256:
+    assert self.expiry_timestamp > block.timestamp, GATE_SEAL_EXPIRED
+    assert self.extensions_remaining > 0, EXTENSIONS_EXHAUSTED
+
+
```
```suggestion
```

Alexvozhak committed:
chore: rename 'extension' to 'prolongation'

Alexvozhak committed:
fix: remarks after the review
- restored committee's right to prolong
- changed terminology from extension to prolongation

Alexvozhak committed:
feat: add maximum prolongations limit to GateSeal contract

Alexvozhak committed:
feat: add maximum prolongation duration limit

Alexvozhak committed:
fix: rewrite tests for strictness
-  fix systemic timestamp handling issue also

Alexvozhak committed:
docs: update README.md
 - clarify dependency diagram
